src/server.h

```diff
@@ -1624,6 +1626,7 @@ struct redisServer {
     list *clients_pending_write; /* There is to write or install handler. */
     list *clients_pending_read;  /* Client has pending read socket buffers. */
     list *slaves, *monitors;    /* List of slaves and MONITORs */
+    dict *pending_slaves;       /* Dict[Slave IP] = replDataBlock with the required data */
```

Comment:
- Currently the dictionary values are the replica's rdb-client. Regarding if the rdb load is done before replica established psync connection, we can keep the rdb client up until the psync conn establishes successfully or until some timeout is reached. However it is a rare case so it sounds a bit like an over kill to me. I think we should allow sync to fail in such cases, as the RDB is probably very small anyway. 

- Lets use Add and Remove instead of Peer and Unpeer, and instead of pending_replicas we can use `slave_waiting_psync`. 

> by replDataBlock you mean replBufBlock, right?

Yep, will fix that

> if we keep it in the current form, i think we should state more clearly that in what state / duration they're listed here (i.e. that they're removed when the main connection is created)

Right I will clarify that

Reply:
> Currently the dictionary values are the replica's rdb-client.

ohh, right (`dictAdd(server.pending_slaves, slave_name, slave);`), so it means this statement is wrong: `Dict[Slave IP] = replDataBlock`

> Regarding if the rdb load is done before replica established psync connection, we can keep the rdb client up until the psync conn establishes successfully or until some timeout is reached. However it is a rare case so it sounds a bit like an over kill to me. I think we should allow sync to fail in such cases, as the RDB is probably very small anyway.

what's rare? the case when the rdb part will complete before the second connection is established? i don't think it's rare.
and what do you mean by allowing it to fail? if it'll be failing psync and resorting to a full one it could be an ok solution, but just failing full sync is wrong IMHO.
in any case, i rather solve all race conditions. either we change the design in a way that there's no race (we wait with the final part of the fork), or we make sure that both cases are handled correctly (and add a test).